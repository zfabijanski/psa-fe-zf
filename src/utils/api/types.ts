export interface IApiError {
  code: string;
  uuid: string;
  status: number;
  exceptions: string[];
  message: string;
}
