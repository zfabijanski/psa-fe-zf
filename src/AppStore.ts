import {
  configureStore,
  ThunkAction,
  ThunkDispatch,
  Action,
  ActionCreator,
  combineReducers,
} from "@reduxjs/toolkit";
import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";

// slices
import { authSlice } from "slices/auth";
import { auxiliaryCalculatorSlice } from "slices/auxiliaryCalculator";
import { bopDropdownListsSlice } from "slices/bopDropdownLists";
import { bopSlice, actions as bopActions } from "slices/bop";
import { businessWithPrudentialSlice } from "slices/businessWithPrudential";
import { calculatorSlice } from "slices/calculator";
import { confirmModalSlice } from "slices/confirmModal";
import { coversConfigSlice } from "slices/coversConfig";
import { dictionariesSlice } from "slices/dictionaries";
import { fullscreenSpinnerSlice } from "slices/fullscreenSpinner";
import { fundRiskProfilesConfigSlice } from "slices/fundRiskProfilesConfig";
import { iddSlice, actions as iddActions } from "slices/idd";
import { librarySlice } from "slices/library";
import { mailSlice } from "slices/mail";
import { meetingsSlice } from "slices/meetings";
import { modalSlice } from "slices/modal";
import { productCoversLimitsSlice } from "slices/productCoversLimits";
import { productsConfigSlice } from "slices/productsConfig";
import { productsSlice } from "slices/products";
import { quarterlyIncomeCalculatorSlice } from "slices/quarterlyIncomeCalculator";
import { translationsSlice } from "slices/translations";
import { illustrationsSlice } from "slices/illustrations";

export const rootReducer = combineReducers({
  auth: authSlice.reducer,
  auxiliaryCalculator: auxiliaryCalculatorSlice.reducer,
  bop: bopSlice.reducer,
  bopDropdownLists: bopDropdownListsSlice.reducer,
  businessWithPrudential: businessWithPrudentialSlice.reducer,
  calculator: calculatorSlice.reducer,
  confirmModal: confirmModalSlice.reducer,
  coversConfig: coversConfigSlice.reducer,
  dictionaries: dictionariesSlice.reducer,
  fullscreenSpinner: fullscreenSpinnerSlice.reducer,
  fundRiskProfilesConfig: fundRiskProfilesConfigSlice.reducer,
  idd: iddSlice.reducer,
  illustrations: illustrationsSlice.reducer,
  library: librarySlice.reducer,
  mail: mailSlice.reducer,
  meetings: meetingsSlice.reducer,
  modal: modalSlice.reducer,
  productCoversLimits: productCoversLimitsSlice.reducer,
  products: productsSlice.reducer,
  productsConfig: productsConfigSlice.reducer,
  quarterlyIncomeCalculator: quarterlyIncomeCalculatorSlice.reducer,
  translations: translationsSlice.reducer,
});

const devToolsActionCreators = {
  auth: authSlice.actions,
  auxiliaryCalculator: auxiliaryCalculatorSlice.actions,
  bop: bopActions,
  bopDropdownLists: bopDropdownListsSlice.actions,
  businessWithPrudential: businessWithPrudentialSlice.actions,
  calculator: calculatorSlice.actions,
  confirmModal: confirmModalSlice.actions,
  coversConfig: coversConfigSlice.actions,
  dictionaries: dictionariesSlice.actions,
  fullscreenSpinner: fullscreenSpinnerSlice.actions,
  fundRiskProfilesConfig: fundRiskProfilesConfigSlice.actions,
  idd: iddActions,
  illustrations: illustrationsSlice.actions,
  library: librarySlice.actions,
  mail: mailSlice.actions,
  meetings: meetingsSlice.actions,
  modal: modalSlice.actions,
  productCoversLimits: productCoversLimitsSlice.actions,
  products: productsSlice.actions,
  productsConfig: productsConfigSlice.actions,
  quarterlyIncomeCalculator: quarterlyIncomeCalculatorSlice.actions,
  translations: translationsSlice.actions,
  // Workaround for the devTools middleware not being able to handle
  // action creators that return a function
} as unknown as Record<string, ActionCreator<unknown>>;

const store = configureStore({
  reducer: rootReducer,
  devTools: {
    name: `PSA Online, timestamp:${Date.now()}`,
    actionCreators: devToolsActionCreators,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

export type RootState = Readonly<ReturnType<typeof rootReducer>>;
export type ThunkResult<T = void> = ThunkAction<
  T,
  RootState,
  undefined,
  Action<any>
>;

export type TAppDispatch = ThunkDispatch<RootState, any, Action>;
export const useAppDispatch = () => useDispatch<TAppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export { configureStore };

export default store;
