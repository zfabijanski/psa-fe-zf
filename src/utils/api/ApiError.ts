import { IApiError } from "./types";

export class ApiError extends Error implements IApiError {
  public readonly code: string;
  public readonly status: number;
  public readonly exceptions: string[];
  public readonly message: string;
  public readonly uuid: string;

  constructor(private apiError: IApiError) {
    super(apiError.message);
    this.code = apiError.code;
    this.status = apiError.status;
    this.exceptions = apiError.exceptions;
    this.message = apiError.message;
    this.uuid = apiError.uuid;
  }
}
