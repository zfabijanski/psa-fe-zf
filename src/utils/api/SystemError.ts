import { AxiosError, AxiosRequestConfig, AxiosResponse } from "axios";

export class SystemError extends Error implements AxiosError {
  public readonly code?: string;
  public readonly config: AxiosRequestConfig;
  public readonly message: string;
  public readonly name: string;
  public readonly request: AxiosRequestConfig;
  public readonly response?: AxiosResponse;
  public readonly stack?: string;
  public readonly isAxiosError: boolean;
  public readonly toJSON: () => object;

  constructor(public systemError: AxiosError) {
    super(systemError.message);
    this.code = systemError.code;
    this.config = systemError.config;
    this.message = systemError.message;
    this.name = systemError.name;
    this.request = systemError.request;
    this.response = systemError.response;
    this.stack = systemError.stack;
    this.isAxiosError = systemError.isAxiosError;
    this.toJSON = systemError.toJSON;
  }
}
