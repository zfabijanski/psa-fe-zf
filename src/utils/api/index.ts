export * from "./Api";
export * from "./ApiError";
export * from "./SystemError";
export * from "./types";
