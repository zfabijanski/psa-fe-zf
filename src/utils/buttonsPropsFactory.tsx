import { IButtonProps } from "components/UI/PruButton/PruButton";
import { FormattedMessage } from "react-intl";

export const getArrowNextConfig = (id?: string): Partial<IButtonProps> => ({
  icon: "arrow-right",
  iconPosition: "right",
  children: <FormattedMessage id={id} />,
});

export const apkButtonConfig: Partial<IButtonProps> = {
  buttonType: "secondary",
  children: <FormattedMessage id="apk.title" />,
};

export const iddButtonConfig: Partial<IButtonProps> = {
  buttonType: "secondary",
  children: <FormattedMessage id="idd.title" />,
};

export const prevButtonConfig: Partial<IButtonProps> = {
  buttonType: "secondary",
  icon: "arrow-left",
  children: <FormattedMessage id="bottomButtonBar.back" />,
};

export const calculateButtonConfig: Partial<IButtonProps> = {
  buttonType: "secondary",
  htmlType: "submit",
  children: <FormattedMessage id="bottomButtonBar.calculate" />,
};
