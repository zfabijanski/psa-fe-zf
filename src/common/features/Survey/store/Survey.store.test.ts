import * as apiUtils from "utils/api";
import {
  CodeTypes,
  IAnswerFull,
  SurveyType,
  TextTransformationCode,
  IQuestionRecord,
  IAnswerTree,
} from "../Survey.types";
import {
  generateSurveyActionType,
  generateSurveyActions,
  generateSurveyStore,
  generateSurveyThunks,
} from "./Survey.store";
import { RootState } from "AppStore";
import { AxiosInstance } from "axios";

jest.mock("utils/api");

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const mockAnswer: IAnswerFull = Object.freeze({
  code: "code",
  dictionary: null,
  isActive: false,
  isApproved: false,
  isEditable: true,
  isMain: true,
  isVisible: true,
  label: "label",
  text_transform_code: TextTransformationCode.TTSE1,
  type: CodeTypes.String,
  value: "123",
});

const mockSubanswer: IAnswerTree = Object.freeze({
  code: "subcode",
  dictionary: null,
  isActive: false,
  isApproved: false,
  isEditable: true,
  isMain: false,
  isVisible: true,
  label: "label",
  text_transform_code: TextTransformationCode.TTSE1,
  type: CodeTypes.String,
  value: "123",
  subanswers: [],
});

const mockQuestion: IQuestionRecord = Object.freeze({
  answer_type_code: "answer_type_code",
  data_type: CodeTypes.Single,
  answerStatus: [mockAnswer],
  cancelStatus: null,
  code: CodeTypes.Single,
  prompt_value: "prompt_value",
  question_value: "question_value",
  tooltip: "tooltip",
  type: "",
});

describe("Survey action type", () => {
  it("should generate bop action type", () => {
    expect(generateSurveyActionType(SurveyType.bop)).toMatchSnapshot();
  });
  it("should generate idd action type", () => {
    expect(generateSurveyActionType(SurveyType.idd)).toMatchSnapshot();
  });
});

describe("Survey actions", () => {
  it("should generate bop actions", () => {
    expect(generateSurveyActions(SurveyType.bop)).toMatchSnapshot();
  });
  it("should generate idd actions", () => {
    expect(generateSurveyActions(SurveyType.idd)).toMatchSnapshot();
  });
});

describe("Survey store", () => {
  const actionType = generateSurveyActionType(SurveyType.bop);
  let actions = generateSurveyActions(SurveyType.bop);
  let store = generateSurveyStore(actionType);

  beforeEach(() => {
    actions = generateSurveyActions(SurveyType.bop);
    store = generateSurveyStore(actionType);
  });

  it("Handles toggleAnswerRefusal", () => {
    let state: typeof store.initialState = {
      ...store.initialState,
      questions: [
        {
          data_type: CodeTypes.Single,
          answer_type_code: "answer_type_code",
          question_value: "question_value",
          answerStatus: [],
          cancelStatus: {
            checked: false,
            id: "0",
            label: "label",
            showModal: false,
          },
          prompt_value: "prompt_value",
          tooltip: "tooltip",
          code: "code",
          type: "",
        },
      ],
    };
    expect(store.reducer(state, actions.toggleAnswerRefusal())).toEqual({
      ...store.initialState,
      questions: [
        {
          answerStatus: [],
          answer_type_code: "answer_type_code",
          cancelStatus: {
            checked: true,
            id: "0",
            label: "label",
            showModal: false,
          },
          code: "code",
          data_type: "SINGLE",
          prompt_value: "prompt_value",
          question_value: "question_value",
          tooltip: "tooltip",
          type: "",
        },
      ],
    });
  });

  it("Handles resetAnswerValues", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [
            {
              data_type: CodeTypes.Single,
              answer_type_code: "answer_type_code",
              question_value: "question_value",
              answerStatus: [
                {
                  ...mockAnswer,
                  value: "123",
                },
              ],
              cancelStatus: null,
              prompt_value: "prompt_value",
              tooltip: "tooltip",
              code: "code",
              type: "",
            },
          ],
        },
        actions.resetAnswers()
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          data_type: CodeTypes.Single,
          answer_type_code: "answer_type_code",
          question_value: "question_value",
          answerStatus: [
            {
              ...mockAnswer,
              value: "",
            },
          ],
          cancelStatus: null,
          prompt_value: "prompt_value",
          tooltip: "tooltip",
          code: "code",
          type: "",
        },
      ],
    });
  });

  it("Handles resetQuestionnaire", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          currentStep: 99,
        },
        actions.resetQuestionnaire()
      )
    ).toEqual({
      questions: [],
      adequacy_id: null,
      isQuestionLoading: false,
      isQuestionDeleting: false,
      isDatePickerOpened: false,
      allSteps: 0,
      currentStep: 0,
      type: "",
    });
  });

  it("Handles updateString", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [mockQuestion],
        },
        actions.updateStringAnswer("code", "new-value-01")
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          ...mockQuestion,
          answerStatus: [
            {
              ...mockAnswer,
              isActive: true,
              value: "new-value-01",
            },
          ],
        },
      ],
    });
  });

  it("Handles updateMultiSelect", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [mockQuestion],
        },
        actions.updateMultiSelectAnswer("code")
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          ...mockQuestion,
          answerStatus: [
            {
              ...mockAnswer,
              isActive: true,
            },
          ],
        },
      ],
    });
  });

  it("Handles updateSingleSelect", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [mockQuestion],
        },
        actions.updateSingleSelectAnswer("code")
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          ...mockQuestion,
          answerStatus: [
            {
              ...mockAnswer,
              isActive: true,
            },
          ],
        },
      ],
    });
  });

  it("Handles UpdateNestedSingleSelect", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [
            {
              ...mockQuestion,
              answerStatus: [
                mockAnswer,
                {
                  ...mockAnswer,
                  code: "333",
                  isActive: true,
                  subanswers: [
                    mockSubanswer,
                    {
                      ...mockSubanswer,
                      code: "234",
                    },
                  ],
                },
              ],
            },
          ],
        },
        actions.updateNestedSingleSelectAnswer("subcode", ["333"])
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          ...mockQuestion,
          answerStatus: [
            mockAnswer,
            {
              ...mockAnswer,
              code: "333",
              isActive: true,
              subanswers: [
                {
                  ...mockSubanswer,
                  isActive: true,
                },
                {
                  ...mockSubanswer,
                  code: "234",
                },
              ],
            },
          ],
        },
      ],
    });
  });

  it("Handles AddNextQuestion", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [mockQuestion],
        },
        actions.addNextQuestion(mockQuestion)
      )
    ).toEqual({
      ...store.initialState,
      questions: [mockQuestion, mockQuestion],
    });
  });

  it("Handles SetAdequacyID", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
        },
        actions.setAdequacyID(123)
      )
    ).toEqual({
      ...store.initialState,
      adequacy_id: 123,
    });
  });

  it("Handles SetSpinner", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
        },
        actions.setSpinner(true)
      )
    ).toEqual({
      ...store.initialState,
      isQuestionLoading: true,
    });
  });

  it("Handles SetQuestionDeletingSpinner", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
        },
        actions.setQuestionDeletingSpinner(true)
      )
    ).toEqual({
      ...store.initialState,
      isQuestionDeleting: true,
    });
  });

  it("Handles SetQuestionnaireProgress", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
        },
        actions.setQuestionnaireProgress(10, 14)
      )
    ).toEqual({
      ...store.initialState,
      allSteps: 14,
      currentStep: 10,
    });
  });

  it("Handles MarkAsApproved", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [
            {
              ...mockQuestion,
              answerStatus: [
                { ...mockAnswer, isActive: true, isApproved: false },
                {
                  ...mockAnswer,
                  isActive: false,
                  isApproved: false,
                  subanswers: [
                    {
                      ...mockSubanswer,
                      isActive: true,
                      isApproved: false,
                    },
                  ],
                },
              ],
            },
          ],
        },
        actions.markAsApproved()
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          ...mockQuestion,
          answerStatus: [
            {
              ...mockAnswer,
              isActive: true,
              isApproved: true,
            },
            {
              ...mockAnswer,
              isActive: false,
              isApproved: false,
              subanswers: [
                {
                  ...mockSubanswer,
                  isActive: true,
                  isApproved: true,
                },
              ],
            },
          ],
        },
      ],
    });
  });

  it("Handles DeleteApproval", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [
            {
              ...mockQuestion,
              answerStatus: [
                {
                  ...mockAnswer,
                  isActive: true,
                  isApproved: true,
                  subanswers: [
                    {
                      ...mockSubanswer,
                      isApproved: true,
                      isActive: true,
                    },
                  ],
                },
                mockAnswer,
              ],
            },
          ],
        },
        actions.deleteApproval()
      )
    ).toEqual({
      ...store.initialState,
      questions: [
        {
          ...mockQuestion,
          answerStatus: [
            {
              ...mockAnswer,
              isActive: true,
              isApproved: false,
              subanswers: [
                {
                  ...mockSubanswer,
                  isApproved: false,
                  isActive: true,
                },
              ],
            },
            mockAnswer,
          ],
        },
      ],
    });
  });

  it("Handles DeleteLastQuestion", () => {
    expect(
      store.reducer(
        {
          ...store.initialState,
          questions: [mockQuestion, mockQuestion],
        },
        actions.deleteLastQuestion(10, 14)
      )
    ).toEqual({
      ...store.initialState,
      currentStep: 10,
      allSteps: 14,
      questions: [mockQuestion],
    });
  });
});

describe("Survey thunks", () => {
  let actions = generateSurveyActions(SurveyType.bop);

  beforeEach(() => {
    actions = generateSurveyActions(SurveyType.bop);
  });

  it("Handles single select click", async () => {
    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn(
      () =>
        ({
          [SurveyType.bop]: {
            questions: [mockQuestion],
          },
        } as RootState)
    );
    const mockGetNextQuestion = jest.fn();
    thunks.getNextQuestion = mockGetNextQuestion();

    thunks.singleSelectClickHandler("code")(dispatch, getState, undefined);

    expect(dispatch).toBeCalledWith({
      payload: { answerCode: "code" },
      type: "@@bop/updateSingleSelect",
    });
  });

  it("Handles stop questionnaire", async () => {
    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn(
      () =>
        ({
          [SurveyType.bop]: {
            questions: [mockQuestion],
          },
        } as RootState)
    );

    thunks.stopQuestionnaire()(dispatch, getState, undefined);

    expect(dispatch).toBeCalledWith({
      payload: {
        modalButtons: {
          cancel: {
            onClick: expect.any(Function),
            textTrKey: "confirmWindow.back",
          },
          confirm: {
            onClick: expect.any(Function),
            textTrKey: "confirmWindow.stopAdequacy",
          },
        },
        modalContentTrKey: "bop.confirm.denyQuestion",
        modalIllustration: "alert-double-triangle",
        modalTitleTrKey: "idd.confirm.denyTitle",
        modalType: "danger",
      },
      type: "modal/showModal",
    });
  });

  it("Should handle refuse answer", () => {
    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn(
      () =>
        ({
          [SurveyType.bop]: {
            questions: [mockQuestion],
          },
        } as RootState)
    );

    thunks.refuseAnswer()(dispatch, getState, undefined);

    expect(dispatch).toBeCalledWith({ type: "@@bop/toggleAnswerRefusal" });
    expect(dispatch).toBeCalledWith({ type: "@@bop/resetAnswerValues" });
  });

  it("Should handle start questionnaire", async () => {
    (apiUtils as jest.Mocked<{ api: Partial<AxiosInstance> }>).api = {
      post: jest.fn().mockReturnValue(
        Promise.resolve({
          adequacy_id: "123",
          step_no: 1,
          step_all_no: 4,
          questionnaire: [mockQuestion],
        })
      ),
    };

    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn(
      () =>
        ({
          [SurveyType.bop]: {
            questions: [mockQuestion],
          },
        } as RootState)
    );

    thunks.startQuestionnaire()(dispatch, getState, undefined);
    await sleep(1);

    expect(dispatch).toBeCalledWith({
      payload: { id: "123" },
      type: "@@bop/setAdequacyId",
    });
    expect(dispatch).toBeCalledWith({
      payload: { allSteps: 4, currentStep: 1 },
      type: "@@bop/setQuestionnaireProgress",
    });
  });

  it("Should reset idd status when starting questionnaire", async () => {
    (apiUtils as jest.Mocked<{ api: Partial<AxiosInstance> }>).api = {
      post: jest.fn().mockReturnValue(
        Promise.resolve({
          adequacy_id: "123",
          step_no: 1,
          step_all_no: 4,
          questionnaire: [mockQuestion],
        })
      ),
    };

    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.idd,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn(
      () =>
        ({
          [SurveyType.idd]: {
            questions: [mockQuestion],
          },
        } as RootState)
    );

    thunks.startQuestionnaire()(dispatch, getState, undefined);
    await sleep(1);

    expect(dispatch).toBeCalledWith({
      payload: { id: "123" },
      type: "@@bop/setAdequacyId",
    });
    expect(dispatch).toBeCalledWith({
      type: "meetings/setCurrentMeetingHaveIdd",
      payload: false,
    });
    expect(dispatch).toBeCalledWith({
      payload: { allSteps: 4, currentStep: 1 },
      type: "@@bop/setQuestionnaireProgress",
    });
  });

  it("Should handle next question", async () => {
    (apiUtils as jest.Mocked<{ api: Partial<AxiosInstance> }>).api = {
      post: jest.fn().mockReturnValue(
        Promise.resolve({
          adequacy_id: "123",
          step_no: 2,
          step_all_no: 4,
          questionnaire: [mockQuestion],
        })
      ),
    };

    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn().mockReturnValue({
      [SurveyType.bop]: {
        adequacyId: "123",
        questions: [mockQuestion],
      },
      fullscreenSpinner: { active: false },
    });

    thunks.getNextQuestion()(dispatch, getState, undefined);
    await sleep(1);

    expect(dispatch).toBeCalledWith({
      payload: { isQuestionLoading: true },
      type: "@@bop/setSpinner",
    });
  });

  it("Should handle next question with error", async () => {
    (apiUtils as jest.Mocked<{ api: Partial<AxiosInstance> }>).api = {
      post: jest.fn().mockReturnValue(Promise.reject()),
    };

    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn().mockReturnValue({
      [SurveyType.bop]: {
        adequacyId: "123",
        questions: [mockQuestion],
      },
      fullscreenSpinner: { active: false },
    });

    thunks.getNextQuestion()(dispatch, getState, undefined);
    await sleep(1);

    expect(dispatch).toBeCalledWith({
      payload: { isQuestionLoading: true },
      type: "@@bop/setSpinner",
    });
  });

  it('Handles "next question" click for MultiGroup data type in IDD', async () => {
    actions = generateSurveyActions(SurveyType.idd);

    (apiUtils as jest.Mocked<{ api: Partial<AxiosInstance> }>).api = {
      post: jest.fn().mockReturnValue(
        Promise.resolve({
          questionnaire_end: "Y",
        })
      ),
    };
    const thunks = generateSurveyThunks({
      actions,
      type: SurveyType.idd,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    const dispatch = jest.fn();
    const getState = jest.fn(
      () =>
        ({
          [SurveyType.idd]: {
            questions: [
              {
                ...mockQuestion,
                data_type: CodeTypes.MultiGroup,
              },
            ],
          },
          fullscreenSpinner: { active: false },
        } as RootState)
    );

    thunks.getNextQuestion()(dispatch, getState, undefined);
    await sleep(1);

    expect(dispatch).toBeCalledWith({
      payload: { answerCode: "code", newValue: "123" },
      type: "@@idd/updateString",
    });
    expect(dispatch).toBeCalledWith({
      payload: { isQuestionLoading: true },
      type: "@@idd/setSpinner",
    });
  });

  it("isValid should be false when string type is empty", () => {
    const { isValid } = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    expect(
      isValid([
        {
          ...mockQuestion,
          data_type: CodeTypes.Single,
          answerStatus: [{ ...mockAnswer, value: "" }],
        },
      ])
    ).toBe(false);
  });

  it("isValid should be false when string type is empty and answer has a validation pattern", () => {
    const { isValid } = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    expect(
      isValid([
        {
          ...mockQuestion,
          data_type: CodeTypes.Single,
          answerStatus: [
            {
              ...mockAnswer,
              value: "",
              validationPattern: "^((?!(0))\\d{0,9})$|^(0{0,1})$",
              validationPatternError: "Please enter a valid number",
            },
          ],
        },
      ])
    ).toBe(false);
  });

  it("isValid should be true when string matches validation pattern", () => {
    const { isValid } = generateSurveyThunks({
      actions,
      type: SurveyType.bop,
      handlers: {
        onStopQuestionnaireConfirm: jest.fn(),
        onStopQuestionnaireCancel: jest.fn(),
      },
    });

    expect(
      isValid([
        {
          ...mockQuestion,
          data_type: CodeTypes.Single,
          answerStatus: [
            {
              ...mockAnswer,
              value: "123",
              validationPattern: "^((?!(0))\\d{0,9})$|^(0{0,1})$",
              validationPatternError: "Please enter a valid number",
            },
          ],
        },
      ])
    ).toBe(true);
  });
});
