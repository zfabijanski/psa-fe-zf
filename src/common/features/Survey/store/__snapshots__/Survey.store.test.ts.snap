// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey action type should generate bop action type 1`] = `
Object {
  "AddNextQuestion": "@@bop/addQuestion",
  "DeleteApproval": "@@bop/deleteApproval",
  "DeleteLastQuestion": "@@bop/deleteLastQuestion",
  "MarkAsApproved": "@@bop/markAsApproved",
  "ResetAnswerValues": "@@bop/resetAnswerValues",
  "ResetQuestionnaire": "@@bop/resetQuestionnaire",
  "SetAdequacyID": "@@bop/setAdequacyId",
  "SetQuestionDeletingSpinner": "@@bop/setQuestionDeletingSpinner",
  "SetQuestionnaireProgress": "@@bop/setQuestionnaireProgress",
  "SetSpinner": "@@bop/setSpinner",
  "ToggleAnswerRefusal": "@@bop/toggleAnswerRefusal",
  "UpdateMultiSelect": "@@bop/updateMultiSelect",
  "UpdateNestedMultiSelect": "@@bop/updateNestedMultiSelect",
  "UpdateNestedSingleSelect": "@@bop/updateNestedSingleSelect",
  "UpdateSingleSelect": "@@bop/updateSingleSelect",
  "UpdateString": "@@bop/updateString",
}
`;

exports[`Survey action type should generate idd action type 1`] = `
Object {
  "AddNextQuestion": "@@idd/addQuestion",
  "DeleteApproval": "@@idd/deleteApproval",
  "DeleteLastQuestion": "@@idd/deleteLastQuestion",
  "MarkAsApproved": "@@idd/markAsApproved",
  "ResetAnswerValues": "@@idd/resetAnswerValues",
  "ResetQuestionnaire": "@@idd/resetQuestionnaire",
  "SetAdequacyID": "@@idd/setAdequacyId",
  "SetQuestionDeletingSpinner": "@@idd/setQuestionDeletingSpinner",
  "SetQuestionnaireProgress": "@@idd/setQuestionnaireProgress",
  "SetSpinner": "@@idd/setSpinner",
  "ToggleAnswerRefusal": "@@idd/toggleAnswerRefusal",
  "UpdateMultiSelect": "@@idd/updateMultiSelect",
  "UpdateNestedMultiSelect": "@@idd/updateNestedMultiSelect",
  "UpdateNestedSingleSelect": "@@idd/updateNestedSingleSelect",
  "UpdateSingleSelect": "@@idd/updateSingleSelect",
  "UpdateString": "@@idd/updateString",
}
`;

exports[`Survey actions should generate bop actions 1`] = `
Object {
  "addNextQuestion": [Function],
  "deleteApproval": [Function],
  "deleteLastQuestion": [Function],
  "markAsApproved": [Function],
  "resetAnswers": [Function],
  "resetQuestionnaire": [Function],
  "setAdequacyID": [Function],
  "setQuestionDeletingSpinner": [Function],
  "setQuestionnaireProgress": [Function],
  "setSpinner": [Function],
  "toggleAnswerRefusal": [Function],
  "updateMultiSelectAnswer": [Function],
  "updateNestedMultiSelectAnswer": [Function],
  "updateNestedSingleSelectAnswer": [Function],
  "updateSingleSelectAnswer": [Function],
  "updateStringAnswer": [Function],
}
`;

exports[`Survey actions should generate idd actions 1`] = `
Object {
  "addNextQuestion": [Function],
  "deleteApproval": [Function],
  "deleteLastQuestion": [Function],
  "markAsApproved": [Function],
  "resetAnswers": [Function],
  "resetQuestionnaire": [Function],
  "setAdequacyID": [Function],
  "setQuestionDeletingSpinner": [Function],
  "setQuestionnaireProgress": [Function],
  "setSpinner": [Function],
  "toggleAnswerRefusal": [Function],
  "updateMultiSelectAnswer": [Function],
  "updateNestedMultiSelectAnswer": [Function],
  "updateNestedSingleSelectAnswer": [Function],
  "updateSingleSelectAnswer": [Function],
  "updateStringAnswer": [Function],
}
`;
